python/ray/tune/schedulers/resource_changing_scheduler.py

```diff
@@ -14,130 +16,374 @@
                 )
-            upper_cpu_limit = max(min_cpu, upper_cpu_limit)
+            upper_cpu_limit = max(self.min_cpu, upper_cpu_limit)
 
-        if min_gpu == 0:
+        if self.min_gpu == 0:
             upper_gpu_limit = 0
         else:
-            upper_gpu_limit = math.ceil(total_available_gpus / num_running_trials)
+            upper_gpu_limit = math.ceil(self.total_available_gpus / num_running_trials)
             # Ensure we don't go below per-bundle minimum
             if self.add_bundles:
                 upper_gpu_limit = (
-                    math.ceil(upper_gpu_limit / min_cpu_bundle) * min_gpu_bundle
+                    math.ceil(upper_gpu_limit / self.min_gpu_bundle)
+                    * self.min_gpu_bundle
                 )
-            upper_gpu_limit = max(min_gpu, upper_gpu_limit)
+            upper_gpu_limit = max(self.min_gpu, upper_gpu_limit)
 
-        # Function to check how many CPUs and GPUs a trial is using currently
-        def get_used_cpus_and_gpus(t: Trial):
-            return (
-                t.placement_group_factory.required_resources.get("CPU", 0),
-                t.placement_group_factory.required_resources.get("GPU", 0),
+        return upper_cpu_limit, upper_gpu_limit
+
+    def _modify_upper_limits_with_increase_by_times(self) -> None:
+        if self.increase_by and self.increase_by_times > 0:
+            required_cpus = self.increase_by.get("CPU", 0)
+            required_gpus = self.increase_by.get("GPU", 0)
+            self.upper_cpu_limit = min(
+                self.upper_cpu_limit,
+                self.min_cpu + required_cpus * self.increase_by_times,
+            )
+            self.upper_gpu_limit = min(
+                self.upper_gpu_limit,
+                self.min_gpu + required_gpus * self.increase_by_times,
             )
 
+    def _modify_lower_limits_with_increase_by_times(self) -> None:
+        if self.increase_by:
+            required_cpus = self.increase_by.get("CPU", 0)
+            required_gpus = self.increase_by.get("GPU", 0)
+            if required_cpus and required_gpus:
+                multiplier = min(
+                    self.free_cpus // required_cpus, self.free_gpus // required_gpus
+                )
+            elif required_gpus:
+                multiplier = self.free_gpus // required_gpus
+            else:
+                multiplier = self.free_cpus // required_cpus
+            multiplier = max(multiplier, 0)
+            self.free_cpus = multiplier * required_cpus
+            self.free_gpus = multiplier * required_gpus
+
+    def _get_used_cpus_and_gpus(self, t: Trial):
+        """Function to check how many CPUs and GPUs a trial is using
+        currently"""
+        return (
+            t.placement_group_factory.required_resources.get("CPU", 0),
+            t.placement_group_factory.required_resources.get("GPU", 0),
+        )
+
+    def __call__(
+        self,
+        trial_runner: "trial_runner.TrialRunner",
```

Comment:
Let's not expose more that what is actually needed. How about just `live_trials`? It's already a big pain for Scheduler to take in `trial_runner` through its interface.

Reply:
This PR merely refactors the class. I am open to changing the API once we change the scheduler API.

Reply:
Hmm, but it does promote a somewhat internal class (`_DistributeResources `) to a public interface right?

Reply:
I suppose, but the interface between the resource allocation function and the resources changing scheduler was and is public - and I wouldn't want to change that here.